### PR TITLE
New Module: vmware_datastore_facts - Gather facts from vmware datastores

### PR DESCRIPTION
**Pull request by garbled1**
##### SUMMARY

A module to gather facts from vmware datastores by cluster or datacenter

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py

##### ANSIBLE VERSION

```
ansible 2.5.0 (vmware_datastore_facts a1eaa6c2c6) last updated 2017/09/23 16:01:22 (GMT +000)
  config file = None
  configured module search path = [u'/home/ec2-user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/ec2-user/ansible/lib/ansible
  executable location = /home/ec2-user/ansible/bin/ansible
  python version = 2.7.12 (default, Sep  1 2016, 22:14:00) [GCC 4.8.3 20140911 (Red Hat 4.8.3-9)]

```

##### ADDITIONAL INFORMATION

This module allows you to gather facts from vmware datastores.

```
- name: Gather facts from standalone ESXi server having datacenter as 'ha-datacenter'
  vmware_datastore_facts:
    hostname: 192.168.1.209
    username: administrator@vsphere.local
    password: vmware
    datacenter: ha-datacenter
    validate_certs: no
  delegate_to: localhost
  register: facts
```

### CODE REVIEW

**Comment by jjahns** [comment]
Seems like we should be adding some of this to the module util as opposed to creating it here. Perhaps a list_datastores function that takes a datacenter parameter?

Interested in what everyone else thinks.

File: lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    - PyVmomi
+options:
+   name:
+        description:
+            - Name of a datastore to match
+   datacenter:
+        description:
+            - Datacenter to search for datastores
+            - This is required if cluster is not supplied
+        required: False
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py]: required is by-default False so no need to mention explicitly.

File: lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    returned: always
+    type: dict
+    sample: None
+"""
+
+try:
+    import pyVmomi
+    from pyVmomi import vim
+
+    HAS_PYVMOMI = True
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py]: use ```from ansible.module_utils.vmware import HAS_PYVMOMI```

File: lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+except ImportError:
+    HAS_PYVMOMI = False
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_text
+from ansible.module_utils.vmware import (connect_to_api, vmware_argument_spec,
+                                         get_all_objs)
+
+
+def find_obj(content, vimtype, name, first=True):
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py]: use ```from ansible.module_utils.vmware import find_obj```

File: lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    # Return all matching objects if needed
+    return [obj for obj in obj_list if obj.name == name]
+
+
+class PyVmomiCache(object):
+    """ This class caches references to objects which are requested multiples times but not modified """
+    def __init__(self, content, dc_name=None):
+        self.content = content
+        self.dc_name = dc_name
+        self.networks = {}
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py]: remove un-used variable.

File: lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        self.clusters = {}
+        self.esx_hosts = {}
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py]: remove un-used variable.

File: lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        self.parent_datacenters = {}
+
+    def get_all_objs(self, content, types, confine_to_datacenter=True):
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py]: use ```from ansible.module_utils.vmware import get_all_objs```

File: lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+                    parent_dc = self.get_parent_datacenter(k)
+                    if parent_dc.name != self.dc_name:
+                        objects.pop(k, None)
+            else:
+                # everything else should be a list
+                objects = [x for x in objects if self.get_parent_datacenter(x).name == self.dc_name]
+
+        return objects
+
+    def get_cluster(self, cluster):
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py]: use ```from ansible.module_utils.vmware import get_cluster```

File: lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+        if cluster not in self.clusters:
+            self.clusters[cluster] = find_obj(self.content, [vim.ClusterComputeResource], cluster)
+        return self.clusters[cluster]
+
+    def get_parent_datacenter(self, obj):
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py]: use ```from ansible.module_utils.vmware import get_parent_datacenter```

File: lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        dds['provisioned'] = summary.capacity - summary.freeSpace + summary.uncommitted
+        if module.params['name']:
+            if dds['name'] == module.params['name']:
+                datastores.extend([dds])
+        else:
+            datastores.extend([dds])
+
+    result['datastores'] = datastores
+
+    # VM already exists
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py]: if datastore already exists

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py (hunk 9 above)
Comment (garbled1) [reply lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py]: Thank you for all the comments.  I will work on these and update.

File: lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py (hunk 8 above)
Comment (garbled1) [reply lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py]: I don't see this one.  Are you suggesting I add it over there?

**Comment by garbled1** [comment]
I believe I have addressed all of the comments that were possible for me to fix without modification to the module_utils directly.

**Comment by Akasurde** [comment]
@garbled1 Could you please write integration test for this change ?

**Comment by garbled1** [comment]
I have added integration tests.  Thank you!

**Comment by garbled1** [comment]
Is this waiting on anything?

File: lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py (hunk 7 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py]: I meant `from ansible.module_utils.vmware import find_cluster_by_name`

File: lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py (hunk 8 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_datastore_facts.py]: My bad. Could you please add this to vmware.py

**Review by Akasurde — CHANGES_REQUESTED** [review]

**Comment by Akasurde** [comment]
@garbled1 This looks good to me. Ping me once when you have suggested changes. I will merge it.

**Comment by garbled1** [comment]
Updated and fixed.

**Comment by Akasurde** [comment]
rebuild_merge

**Comment by garbled1** [comment]
did you need me to rebase this one?

**Review by Akasurde — APPROVED** [review]
shipit

**Comment by Akasurde** [comment]
rebuild_merge

**Pull request merged**